sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/CBDE/CbdeHandler.cs

```diff
@@ -0,0 +1,328 @@
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Linq;
+using System;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using SonarAnalyzer.Helpers;
+using System.IO;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.CodeAnalysis.Diagnostics;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Text;
+using System.Threading;
+
+namespace SonarAnalyzer.Rules.CSharp
+{
+    public class CbdeHandler
+    {
+        private readonly Action<String, String, Location, CompilationAnalysisContext> raiseIssue;
+        private readonly Func<CompilationAnalysisContext, bool> shouldRunInContext;
+        private static bool initialized = false;
+
+        private const string cbdeJsonOutputFileName = "cbdeSEout.json";
+
+        private static string cbdeBinaryPath;
+        private string mlirDirectoryRoot;
+        private string mlirDirectoryAssembly;
+        private string cbdeJsonOutputPath;
+
+        protected HashSet<string> csSourceFileNames= new HashSet<string>();
+        protected Dictionary<string, int> fileNameDuplicateNumbering = new Dictionary<string, int>();
+
+        private MemoryStream logStream;
+        private StreamWriter logStreamWriter;
+        private static readonly object logFileLock = new Object();
+        private static readonly object metricsFileLock = new Object();
+        private static readonly object perfFileLock = new Object();
+        private static readonly object staticInitLock = new Object();
+
+        private static readonly string mlirPath =
+            Environment.GetEnvironmentVariable("CIRRUS_WORKING_DIR") ?? // For Cirrus
+            Environment.GetEnvironmentVariable("WORKSPACE") ?? // For Jenkins
+            Path.GetTempPath(); // By default
+        private static readonly string mlirProcessSpecificPath = Path.Combine(mlirPath, $"CBDE_{Process.GetCurrentProcess().Id}");
+        private static readonly string mlirLogFile = Path.Combine(mlirProcessSpecificPath, "cbdeHandler.log");
+        private static readonly string mlirMetricsLogFile = Path.Combine(mlirProcessSpecificPath, "metrics.log");
+        private static readonly string mlirPerfLogFile = Path.Combine(mlirProcessSpecificPath, "performances.log");
+        private static void GlobalLog(string s)
+        {
+            lock (logFileLock)
```

Comment:
should you do this only when it runs in DEBUG mode?
or when some property is turned on (and it should be turned off by default, in PROD)?

Reply:
see [PropertiesHelper](https://github.com/SonarSource/sonar-dotnet/blob/master/sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Helpers/PropertiesHelper.cs#L30) and how it's used

Reply:
in the end we agreed to use an env var, as the usage is mainly internal. It will also be easy to use in case of support diagnostics